### PR TITLE
Remove `Serialise` class from a bunch of places

### PR DESCRIPTION
**Pull request by nc6**
We remove `Serialise` from a variety of places in the consensus layer, in order to allow integration with the ledger layer, which does not use `Serialise`.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Crypto/DSIGN/RSAPSS.hs [hunk 1]
```diff
@@ -42,23 +42,31 @@ instance DSIGNAlgorithm RSAPSSDSIGN where
...
 
-    verifyDSIGN (VerKeyRSAPSSDSIGN vk) a (SigRSAPSSDSIGN sig) =
-        verify defaultPSSParamsSHA1 vk (toBS a) sig
+    verifyDSIGN toEnc (VerKeyRSAPSSDSIGN vk) a (SigRSAPSSDSIGN sig) =
+        verify defaultPSSParamsSHA1 vk (toBS $ toEnc a) sig
 
-toBS :: Serialise a => a -> ByteString
-toBS = toStrict . serialise
+-- toBS :: Serialise a => a -> ByteString
+-- toBS = toStrict . serialise
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Crypto/DSIGN/RSAPSS.hs]: You can remove this

File: ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs [hunk 2]
```diff
@@ -35,38 +38,79 @@ instance DSIGNAlgorithm d => KESAlgorithm (SimpleKES d) where
...
+            Just vk -> verifyDSIGN toEnc vk a sig
 
 deriving instance DSIGNAlgorithm d => Show (VerKeyKES (SimpleKES d))
 deriving instance DSIGNAlgorithm d => Eq (VerKeyKES (SimpleKES d))
 deriving instance DSIGNAlgorithm d => Ord (VerKeyKES (SimpleKES d))
-deriving instance DSIGNAlgorithm d => Serialise (VerKeyKES (SimpleKES d))
+instance DSIGNAlgorithm d => Serialise (VerKeyKES (SimpleKES d)) where
+  encode (VerKeySimpleKES vvks) =
+       Enc.encodeListLen (fromIntegral $ Vec.length vvks)
+    <> Vec.foldl' (<>) mempty (fmap encodeVerKeyDSIGN vvks)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs]: Just wondering: why a strict fold here and not lazy (`foldr`)?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock.hs [hunk 3]
```diff
@@ -362,18 +362,27 @@ deriving instance OuroborosTag p => Show (LedgerState (SimpleBlock p c))
 -- This is needed in order to assign stake to nodes.
 type AddrDist = Map Addr NodeId
 
-instance (BftCrypto c, SimpleBlockCrypto c')
+instance (BftCrypto c, SimpleBlockCrypto c', Serialise (Payload (Bft c) (SimplePreHeader (Bft c) c')))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock.hs]: It still compiles if you remove this new `Serialise (Payload ..)` constraint from this instance and the two instances below.

File: ouroboros-consensus/test-crypto/Test/Crypto/DSIGN.hs [hunk 4]
```diff
@@ -1,10 +1,12 @@
+{-# LANGUAGE FlexibleContexts    #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications    #-}
 
 module Test.Crypto.DSIGN
     ( tests
     ) where
 
+import           Codec.Serialise (Serialise(..))
```
Comment (mrBliss) [inline ouroboros-consensus/test-crypto/Test/Crypto/DSIGN.hs]: Duplicate import: `Serialise` is already imported by the line below:
```haskell
import Ouroboros.Network.Testing.Serialise (Serialise, prop_serialise)
```

File: ouroboros-consensus/test-crypto/Test/Crypto/KES.hs [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 module Test.Crypto.KES
     ( tests
     ) where
 
+import           Codec.Serialise (Serialise(..))
```
Comment (mrBliss) [inline ouroboros-consensus/test-crypto/Test/Crypto/KES.hs]: Same duplicate import

File: ouroboros-consensus/test-crypto/Test/Crypto/VRF.hs [hunk 6]
```diff
@@ -1,10 +1,12 @@
+{-# LANGUAGE FlexibleContexts    #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications    #-}
 
 module Test.Crypto.VRF
     ( tests
     ) where
 
+import           Codec.Serialise (Serialise(encode))
```
Comment (mrBliss) [inline ouroboros-consensus/test-crypto/Test/Crypto/VRF.hs]: Same duplicate import

File: ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs (hunk 2 above)
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs]: No specific reason, I think it's just my default fold...

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs]: Same here :slightly_smiling_face:, unless I think being lazy would be useful. I'm not sure, but maybe in this case it is? It depends on the length of the vector as well. How long is the length of this vector typically?

If it is very long, then being lazy could be beneficial, as we wouldn't need to keep the encoding of each element in memory and just get them on-demand. However, I think `Encoding` already is a `Builder`-like thing that allows efficient `mappend`s.

**Comment by mrBliss** [comment]
Before merging this, this branch can be squashed to two commits: one for the .gitignore and one for the rest.

**Comment by nc6** [comment]
Yep, I intend to do that

File: ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs (hunk 2 above)
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs]: Yes, `Encoding` is already such a thing.

**Pull request merged**